phabricatoremails/models.py

```diff
@@ -0,0 +1,31 @@
+from sqlalchemy import Boolean, Column, Integer, BigInteger, CheckConstraint
+from sqlalchemy.ext.declarative import declarative_base
+
+Base = declarative_base()
+
+
+class QueryPosition(Base):
+    """The current position in the Phabricator feed
+
+    Uses the "chronologicalKey" of feed events to track the current position in the
+    Phabricator feed. There should be one and only one row in the table at a time.
+    """
+
+    __tablename__ = "query_position"
+    id = Column(Boolean, CheckConstraint("id"), primary_key=True, default=True)
+    up_to_key = Column(BigInteger, nullable=False)
+    bap = Column(Integer)
+
+
+class Thread(Base):
+    """The amount of messages in an email thread
+
+    We need to add a unique number to each email thread message (due to GMail), and this
+    table provides that unique number for us by tracking the number of messages in each
+    thread.
+    """
+
+    __tablename__ = "thread"
+    id = Column(Integer, primary_key=True)
+    phabricator_id = Column(Integer, nullable=False, unique=True)
+    thread_size = Column(Integer, nullable=False)
```

Comment:
I agree with earlier comments that docstrings are important for realizing the structure and flow of a code base. However, since docs can become out-of-date, it can be worth balancing how extensive your documentation is versus how much information can be gleaned from the code.

I don't think it makes sense to document every variable or field. In this case, what `thread_size` is matters in the context where it's used. I can add docs here explaining that:
* GMail hides information that's been seen before ([documented in-context here](https://github.com/mozilla-conduit/phabricator-emails/blob/1609266-phab-emails/phabricatoremails/render/templates/html/_macros.html.jinja2#L14-L16))
* Each email has a part near the top to populate GMail/iCloud email previews ([code is here](https://github.com/mozilla-conduit/phabricator-emails/blob/1609266-phab-emails/phabricatoremails/render/templates/html/_macros.html.jinja2#L7))
* To avoid the preview being hidden, we put a unique amount of whitespace after it ([code is here](https://github.com/mozilla-conduit/phabricator-emails/blob/1609266-phab-emails/phabricatoremails/render/templates/html/_macros.html.jinja2#L18))
* To populate this unique number, we track the number of events that have happened for a revision (this is discoverable [by finding](https://github.com/mozilla-conduit/phabricator-emails/blob/1609266-phab-emails/phabricatoremails/render/templates/html/normal/abandoned.html.jinja2#L5) how [`unique_number` is defined](https://github.com/mozilla-conduit/phabricator-emails/blob/1609266-phab-emails/phabricatoremails/render/mailbatch.py#L74). ([it's by `thread.thread_size`](https://github.com/mozilla-conduit/phabricator-emails/blob/1609266-phab-emails/phabricatoremails/render/render.py#L190))

However, this load of information should not be described here in `models.py`. It's better discovered in the code itself.

(I'm not saying my or anyone's code is self-documenting, but I am saying that there is always a line before context is better collected by following code, rather than an explicit definition)

Reply:
I agree there there is a balance :+1:

> No need to thoroughly explain all the dependencies and usage instructions (though that doesn't necessarily hurt)

To be pedantic: there _is_ a cost to over-documenting: there's a maintenance burden to docs, and it's not as easy to keep docs up to date as it is code.

For example, in the example from above:
```
    # Track the number of events that have happened for a revision
    thread_size = Column(Integer, nullable=False)
```
I believe that this specific example comment is providing too much detail.
Let's imagine that we change how our render pipeline determines the thread size - maybe it increases it once per recipient per event, maybe something else. `render.py` is changed, though the database schema isn't modified because it's already sufficient.

Unless the developer making the "thread size" change (or the reviewer) is already aware of these docs, they can be missed. Once the change lands, this documentation comment will be worse than missing docs: it will be _incorrect_ docs.

-----

I think that there's some great uses cases for docs, and some examples are:
* For libraries, every public class/method/variable/enum/etc should be documented, because 3rd party devs need to know how it works
* For any project, some high level docs of how the pieces fit together are useful. Architecture docs, so to speak
* It can be valuable to document classes. Sometimes it's especially valuable (e.g.: common abstractions used within a codebase), sometimes less so
* Adding inline docs to fill in the blanks (explaining the use of a constant, an algorithm anomoly, etc) are great

I think that systematically adding a polished pydoc to every method is overkill. I think that, based on the limited usage and scope of this table, describing each of its columns is adding an unnecessary maintainence burden.

-----

Related: good call on renaming `phabricator_id` to `phabricator_revision_id`, I'll take care of that

Reply:
Documentation was added for the columns